### PR TITLE
Feature/slider setting

### PR DESCRIPTION
**Pull request by Sisha3342**

### CODE REVIEW

File: src/settings/slider-setting/slider-setting.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import {UIPSetting} from '../setting/setting';
+import {attr} from '@exadel/esl/modules/esl-base-element/core';
+import {WARNING_MSG} from '../../utils/warning-msg';
+
+export class UIPSliderSetting extends UIPSetting {
+  public static is = 'uip-slider-setting';
+  /** CSS Class added when setting has inconsistent state. */
+  public static inconsistencyClass = 'inconsistency-marker';
```
Comment (nattallius) [inline src/settings/slider-setting/slider-setting.ts:8]: I don't see any actual uses of this

File: src/settings/slider-setting/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Extends [UIPSetting](src/settings/setting/README.md).
+
+## Description:
+
+Setting behaves like 'range' input with value displayed below. It has the
+following attributes:
+
+- *min* - minimum range value.
+- *max* - maximum range value.
+- *step* - step between range's values.
```
Comment (nattallius) [inline src/settings/slider-setting/README.md:13]: Maybe it's better to mention default values in README?

**Review by nattallius — APPROVED** [review]

File: src/settings/slider-setting/slider-setting.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  protected updateSliderValue() {
+    this.$fieldValue.textContent = `Value: ${this.$field.value}`;
+  }
+
+  protected getDisplayedValue(): string {
+    return this.$field.value;
+  }
+
+  protected setValue(value: string | null): void {
+    this.$field.value = value || '0';
```
Comment (nattallius) [inline src/settings/slider-setting/slider-setting.ts:62]: Maybe it's better to set `this.min` instead of hardcoded `'0'` ?

**Review by AlexanderBazukevich — APPROVED** [review]

File: src/settings/slider-setting/slider-setting.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  public static inconsistencyClass = 'inconsistency-marker';
+
+  /** Setting's visible name. */
+  @attr({defaultValue: ''}) public label: string;
+  /** Minimum range value. */
+  @attr({defaultValue: '0'}) public min: string;
+  /** Maximum range value. */
+  @attr({defaultValue: '100'}) public max: string;
+  /** Step for range. */
+  @attr({defaultValue: '1'}) public step: string;
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:17]: It sounds like an '0' will be a more expectable value here

File: src/settings/slider-setting/slider-setting.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  /** Range input to change setting's value. */
+  protected $field: HTMLInputElement;
+  /** Range value displayed below. */
+  protected $fieldValue: HTMLDivElement;
+
+  protected connectedCallback() {
+    super.connectedCallback();
+    if (this.$field) return;
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:25]: I see a little bit of inconsistency here, we assume that the this.$field to be empty but at the same time
- we clear everything inside setting on the 30th line
- and we, basically, have no disconnected callback to make the field empty...

So as I see right now this custom element can not properly control its lifecycle.

File: src/settings/slider-setting/slider-setting.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    this.initSlider();
+    this.initSliderValue();
+    const label = document.createElement('label');
+    label.innerText = this.label;
+    label.appendChild(this.$field);
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:31]: Propose to move it to the separate method

File: src/settings/slider-setting/slider-setting.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    this.innerHTML = '';
+    this.append(label);
+    this.append(this.$fieldValue);
+  }
+
+  protected initSlider() {
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:38]: You can create memoized getter to simplify and optimize things.

Somewhere in the class: 
```
@memoize
protected get $field() {
   const $field = document.createElement('input');
   $field.type = 'range';
   return $field; 
} 

@memoize
protected get $label() {
   const $label = document.createElement('label');
   $label.appendChild(this.$field);
   return $label; 
} 

protected updateConfiguration() {
    this.$field.min = this.min;
    this.$field.max = this.max;
    this.$field.step = this.step;
    // Better wrap the text to the span to easily control it or render it trough ::before { content: attr(label); }
    // this.$label.innerText = this.label;
}

```

In Constructor:
```
this.updateConfiguration();
this.innerHTML = '';
this.append(this.$label);
this.append(this.$fieldValue);
```

File: src/settings/slider-setting/slider-setting.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    this.$field = document.createElement('input');
+    this.$field.type = 'range';
+    this.$field.min = this.min;
+    this.$field.max = this.max;
+    this.$field.step = this.step;
+    this.$field.name = this.label;
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:44]: 'name' should be defined for the form fields, not sure that our field should relates to form

File: src/settings/slider-setting/slider-setting.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  protected initSliderValue() {
+    this.$fieldValue = document.createElement('div');
+    this.$fieldValue.classList.add('slider-value');
+    this.$field.addEventListener('input', () => this.updateSliderValue());
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:50]: There is no unsubscriber for that listener

**Review by ala-n — CHANGES_REQUESTED** [review]

File: src/settings/slider-setting/slider-setting.ts [hunk 10]
```diff
@@ -1,70 +1,80 @@
...
+  @memoize()
+  protected get $fieldValue(): HTMLDivElement {
+    const $fieldValue = document.createElement('div');
+    $fieldValue.classList.add('slider-value');
+    return $fieldValue;
   }
 
+  protected updateSliderValue = () => {
+    this.$fieldValue.textContent = `Value: ${this.$field.value}`;
+  };
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:59]:
```suggestion
  @bind
  protected updateSliderValue() {
    this.$fieldValue.textContent = `Value: ${this.$field.value}`;
  }
```

File: src/settings/slider-setting/slider-setting.ts [hunk 11]
```diff
@@ -1,70 +1,80 @@
...
-    this.$field.value = '0';
+    this.$field.value = this.min;
     this.$fieldValue.textContent = msg;
   }
+
+  protected disconnectedCallback() {
+    this.$field.removeEventListener('input', this.updateSliderValue);
+    this.innerHTML = '';
+    super.disconnectedCallback();
+  }
```
Comment (ala-n) [inline src/settings/slider-setting/slider-setting.ts:79]: Super cosmetic: place it near connected callback for consistency with other components.

**Review by ala-n — APPROVED** [review]

File: src/settings/slider-setting/slider-setting.ts (hunk 4 above)
Comment (ala-n) [reply src/settings/slider-setting/slider-setting.ts]: Just in case: it sounds like I miss rid that it is a step but not a current value.

**Pull request merged**